src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -288,13 +307,33 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
         let mut hir_id = body_id;
         while let Some(node) = self.tcx.hir().find(hir_id) {
             match node {
+                hir::Node::Item(hir::Item {
+                    ident,
+                    kind: hir::ItemKind::Trait(_, _, generics, bounds, _),
+                    ..
+                }) if param_ty && self_ty == self.tcx.types.self_param => {
```

Comment:
I don't think you need the `param_ty &&` part of this test, it's redundant, isn't it? Maybe move it to an `assert!`?